Enable CryptoKit for all releases

nmahendru opened the pull request:
#### 0bcf166a295a09d48406bc93c5cbb5371877586b
<pre>
Enable CryptoKit for all releases
<a href="https://bugs.webkit.org/show_bug.cgi?id=277432">https://bugs.webkit.org/show_bug.cgi?id=277432</a>
<a href="https://rdar.apple.com/132913336">rdar://132913336</a>

Reviewed by Matthew Finkel.

Changing the enablement logic to enable by default and only skip
for platforms we don&apos;t want to. This will also fix the recurring problem of enabling
it in every release.
Also refactor the hardcoded enablement to constexpr instead of individual values as it makes
it easier to revert, just in case.

* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
(WebCore::AuthenticatorAttestationResponse::getPublicKey const):
* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp:
(TestWebKitAPI::TEST(CtapPinTest, TestSetPinRequest)):
(TestWebKitAPI::TEST(CtapPinTest, TestKeyAgreementResponse)):
(TestWebKitAPI::TEST(CtapPinTest, TestTokenRequest)):
* LayoutTests/platform/ios-17/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-ventura/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/281832@main">https://commits.webkit.org/281832@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef4f5f0fe2bdb48d27d5fc5fdbb79b55cd78645e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60673 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40032 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13249 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64604 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11220 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62803 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47708 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11495 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49065 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7789 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62707 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37281 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29893 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33970 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9786 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10133 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55823 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10086 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66333 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4617 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9929 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56434 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4638 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52530 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56612 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3821 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9199 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/35837 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/36919 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38012 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36664 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebCore/Configurations/WebCore.xcconfig:
```diff
@@ -33,23 +33,12 @@ WK_USE_APPLEJPEGXL_MACOS_BEFORE_1400 = YES;
-WK_HAVE_SWIFT_CPP_INTEROP = HAVE_SWIFT_CPP_INTEROP=0;
-WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx15*] = ;
-WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14*] = ;
+WK_HAVE_SWIFT_CPP_INTEROP = ;
```
nit: You don't need this line, build settings are empty by default

emw-apple commented on Source/WebCore/PAL/Configurations/PAL.xcconfig:
```diff
@@ -24,22 +24,12 @@
-WK_HAVE_SWIFT_CPP_INTEROP = HAVE_SWIFT_CPP_INTEROP=0;
-WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx15*] = ;
-WK_HAVE_SWIFT_CPP_INTEROP[sdk=macosx14*] = ;
+WK_HAVE_SWIFT_CPP_INTEROP = ;
```
ditto

nmahendru replied:
done

emw-apple commented on Source/WebCore/PAL/Configurations/PAL.xcconfig:
```diff
@@ -105,22 +95,12 @@ OTHER_SWIFT_FLAGS_IOS_SINCE_17 = -Xcc -std=c++20 -I $(SRCROOT)/pal -I $(SRCROOT)
-EXCLUDED_SOURCE_FILE_NAMES = CryptoKitShim.swift UnsafeOverlays.swift;
-EXCLUDED_SOURCE_FILE_NAMES[sdk=macosx14*] = ;
-EXCLUDED_SOURCE_FILE_NAMES[sdk=macosx15*] = ;
+EXCLUDED_SOURCE_FILE_NAMES = ;
```
ditto

nmahendru replied:
done

emw-apple commented on Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
```diff
@@ -36,6 +36,12 @@
+static constexpr auto useCryptoKit = UseCryptoKit::Yes;
+#else
+static constexpr auto useCryptoKit = UseCryptoKit::No;
+#endif
```
I think you should refactor the WebCore interfaces that take a `useCryptoKit` parameter: remove the parameter entirely, and use `#if` conditions inside each function to decide whether to call into CryptoKit or CommonCrypto.

As is, the code suggests that it's possible for to call into CC on modern platforms by passing the wrong parameter.

nmahendru replied:
it turns out.. All the APIs already have the same check inside them so we don't need the `#if HAVE(SWIFT_CPP_INTEROP)` here and I have removed it..

emw-apple commented on Source/WebCore/Modules/webauthn/fido/Pin.cpp:
```diff
@@ -51,7 +51,9 @@
 using namespace WebCore;
+
 static constexpr auto useCryptoKit = UseCryptoKit::Yes;
+
```
nit: unnecessary whitespace

emw-apple commented on Source/WebCore/PAL/Configurations/PAL.xcconfig:
```diff
@@ -24,22 +24,11 @@
 #include "../../Configurations/FeatureDefines.xcconfig"
 #include "../../../../Configurations/Version.xcconfig"
 
-// FIXME: remove the interop setting here when rdar://124492036 is resolved.
```
ditto above

emw-apple commented on Source/WebCore/PAL/Configurations/PAL.xcconfig:
```diff
@@ -105,22 +94,11 @@ OTHER_SWIFT_FLAGS_IOS_SINCE_17 = -Xcc -std=c++20 -I $(SRCROOT)/pal -I $(SRCROOT)
 OTHER_SWIFT_FLAGS_WATCHOS_SINCE_10 = -Xcc -std=c++20 -I $(SRCROOT)/pal -I $(SRCROOT)/PAL/pal;
 OTHER_SWIFT_FLAGS_TVOS_SINCE_17 = -Xcc -std=c++20 -I $(SRCROOT)/pal -I $(SRCROOT)/PAL/pal;
 
-// FIXME: remove the interop setting here when rdar://124492036 is resolved.
```
ditto

emw-apple commented on Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp:
```diff
@@ -49,6 +49,8 @@ using namespace cbor;
 using namespace fido::pin;
 
+static constexpr auto useCryptoKit = WebCore::UseCryptoKit::Yes;
+
```
ditto above

emw-apple commented on Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
```diff
@@ -36,6 +36,8 @@
 namespace WebCore {
 
+static constexpr auto useCryptoKit = UseCryptoKit::Yes;
+
```
Why make this change?

I think you should really remove the `useCryptoKit` parameter altogether so that behavior cannot be altered at runtime, like I said before. I'm okay with you doing that as a follow up, but you shouldn't needlessly pull this out into a static variable.

nmahendru replied:
I am keeping this as this was the hook i used to develop this and it's also my worst case fallback switch in case we find an issue before shipping.

sysrqb replied:
Can you just reverting these changes so the patch is simpler?

nmahendru committed:
Enable CryptoKit for all releases
https://bugs.webkit.org/show_bug.cgi?id=277432
rdar://132913336

Reviewed by Matthew Finkel.

Changing the enablement logic to enable by default and only skip
for platforms we don't want to. This will also fix the recurring problem of enabling
it in every release.
Also refactor the hardcoded enablement to constexpr instead of individual values as it makes
it easier to revert, just in case.

* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Modules/webauthn/AuthenticatorAttestationResponse.cpp:
(WebCore::AuthenticatorAttestationResponse::getPublicKey const):
* Source/WebCore/Modules/webauthn/fido/Pin.cpp:
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Tools/TestWebKitAPI/Tests/WebCore/CtapPinTest.cpp:
(TestWebKitAPI::TEST(CtapPinTest, TestSetPinRequest)):
(TestWebKitAPI::TEST(CtapPinTest, TestKeyAgreementResponse)):
(TestWebKitAPI::TEST(CtapPinTest, TestTokenRequest)):
* LayoutTests/platform/ios-17/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-ventura/TestExpectations:

Canonical link: https://commits.webkit.org/281832@main
